### PR TITLE
Clarify Java scaffold

### PR DESCRIPTION
**Pull request by ThomasCollignon**

### CODE REVIEW

File: scaffolds/java/README.md [hunk 1]
```diff
@@ -1,35 +1,47 @@
...
 
-There are several options. The easiest way is to work with a docker image ready
-with all the tools you need.
+```
+./gradle run
+```
 
-The only downside is that you need to run gradle through Docker. To make it easy, \
-we wrote a script `gradle` wrapping the `docker` command and acting as gradle.
+## Option 2: Install Java and Gradle 4
```
Comment (jbruggem) [inline scaffolds/java/README.md:23]: We should definitely update to gradle 7 :). Would you be OK to include the update here ?

File: scaffolds/java/scripts/test [hunk 2]
```diff
@@ -2,6 +2,6 @@
 
 cd $(dirname $0)/..
 
-./gradle check
+./gradle test
```
Comment (jbruggem) [inline scaffolds/java/scripts/test:5]: What's the purpose of this change ?

File: scaffolds/java/README.md [hunk 3]
```diff
@@ -1,35 +1,47 @@
 # Java Scaffold for Code Retreat
 
-## Option 1: use docker wrapper scripts
+## Option 1: use Docker wrapper scripts
 
-See `scripts/` folder.
+This scaffold comes with pre-configured Docker environment.
```
Comment (jbruggem) [inline scaffolds/java/README.md:5]:
```suggestion
This scaffold comes a with pre-configured Docker environment.
```

File: scaffolds/java/README.md [hunk 4]
```diff
@@ -1,35 +1,47 @@
...
 
-## Install Java and Gradle
+### Run the tests
+May need to pull the image (600MB) the 1st time.
```
Comment (jbruggem) [inline scaffolds/java/README.md:8]:
```suggestion
You may need to pull the image (several hundred megabytes) the first time.
```

File: scaffolds/java/README.md [hunk 5]
```diff
@@ -1,35 +1,47 @@
...
-There are several options. The easiest way is to work with a docker image ready
-with all the tools you need.
+```
+./gradle run
+```
 
-The only downside is that you need to run gradle through Docker. To make it easy, \
-we wrote a script `gradle` wrapping the `docker` command and acting as gradle.
+## Option 2: Install Java and Gradle 4
+TCR/BTCR scripts are not available because they use Docker.
```
Comment (jbruggem) [inline scaffolds/java/README.md:24]:
```suggestion
With this options, TCR/BTCR scripts will not be directly available because they use `./gradle` 
(which calls Docker). If you wish to use TCR without docker, you can change either `./gradle` 
or the scripts in `./scripts/` so that it calls gradle directly without wrapping in a docker.
```

**Review by jbruggem — CHANGES_REQUESTED** [review]
Thanks for the PR :+1:

File: scaffolds/java/build.gradle [hunk 6]
```diff
@@ -12,8 +10,6 @@ dependencies {
 }
 
 test {
-  testLogging {
-      events "passed", "skipped", "failed", "standardOut", "standardError"
-  }
+  testLogging.showStandardStreams = true
 }
```
[the remark this answers is not in the document]
Comment (ThomasCollignon) [reply scaffolds/java/build.gradle:12]: will do

File: scaffolds/java/README.md (hunk 3 above)
Comment (ThomasCollignon) [reply scaffolds/java/README.md:5]: ? typo ?

File: scaffolds/java/README.md (hunk 5 above)
Comment (ThomasCollignon) [reply scaffolds/java/README.md]: will do

File: scaffolds/java/scripts/test (hunk 2 above)
Comment (ThomasCollignon) [reply scaffolds/java/scripts/test]: To my knowledge, `check` is not commonly used, while `test` is widely used.
Furthermore, `check` makes sense when adding some extra task to the verification process, which is not the case of this scaffold.
[source](https://stackoverflow.com/a/50105980/3790512)

Comment (jbruggem) [reply scaffolds/java/scripts/test]: ok !

File: scaffolds/java/README.md [hunk 7]
```diff
@@ -1,35 +1,53 @@
...
 
-## Run the tests
+## Option 2: use a locally installed Java environment
+This option allows faster Gradle runs thanks to the Gradle daemon.
+Java is required, for ex. 
+[OpenJDK 11 tar.gz](https://download.java.net/java/GA/jdk11/9/GPL/openjdk-11.0.2_linux-x64_bin.tar.gz).
 
-```Bash
-./gradle test
+With this options, TCR/BTCR scripts will not be directly available because they use `./gradle.sh`
```
Comment (jbruggem) [inline scaffolds/java/README.md:28]:
```suggestion
With this option, TCR/BTCR scripts will not be directly available because they use `./gradle.sh`
```

File: scaffolds/java/scripts/build [hunk 8]
```diff
@@ -2,6 +2,6 @@
 
 cd $(dirname $0)/..
 
-./gradle build -x test
+./gradle.sh assemble # or "./gradlew assemble" if you don't want to use Docker
```
Comment (jbruggem) [inline scaffolds/java/scripts/build:5]: Your comment makes me think that we could/should make `./gradle` (AKA `./gradle.sh` or `./gradle-env`) capable of switching automatically between the different options:

```bash
#!/bin/bash

cd "$(dirname "$0")" || exit 1

if [[ -f ./gradlew && -x ./gradlew  ]]; then
  ./gradlew "$@"
elif which gradle > /dev/null; then
  gradle "$@"
else
  docker run --rm \
    -u $(id -u):$(id -g) \
    -v "$(pwd)":/home/gradle/project:rw \
    -w /home/gradle/project \
    gradle:7.4.2-jdk11-alpine gradle "$@"
fi
```

WDYT ?

**Review by jbruggem — CHANGES_REQUESTED** [review]
Looking good so far ! A few suggestions/nit-picking :-)

File: scaffolds/java/gradle/wrapper/gradle-wrapper.properties [hunk 9]
```diff
@@ -0,0 +1,5 @@
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.4.2-bin.zip
```
[the remark this answers is not in the document]
Comment (ThomasCollignon) [reply scaffolds/java/gradle/wrapper/gradle-wrapper.properties:1]: This one is tricky, and seems to divide the community, see [why-should-the-gradle-wrapper-be-committed-to-vcs](https://stackoverflow.com/questions/20348451/why-should-the-gradle-wrapper-be-committed-to-vcs).

**Option 1 - wrapper binary**
If we commit the wrapper, the drawback is that you have binaries in your repo.

**Option 2 - wrapper task**
To remove this drawback, we can add `wrapper {gradleVersion = '7.4.2'}` in `build.gradle`. Then the developer will have to execute `gradle wrapper` and use `./gradlew` for subsequent Gradle tasks. Which means that it requires to have a Gradle version already installed, and if it doesn't match the syntax in `build.gradle`, then he's screwed.

**Option 3 - Gradle as a prerequisite**
We can also simply add in the Readme that having Gradle 5+ installed is a prerequisite in case of no Docker, as it is for the JDK.

**Option 4 - installer script**
Because option 3 is tedious, we could provide a script that downloads and install Gradle (and, why not, Java). But isn't it the very reason of using Docker ? (and I'm not good at Shell scripting)

Personally I would go for option 1. Binaries in a repo is a problem in case it is regularly updated, which may cause issues in term of disk space use, repo cloning speed, and maybe Git perfs. In our case I don't see why we would need to upgrade to Gradle 7.4.3, then 7.5.x, then 7.6.x etc., because our `build.gradle` is concise and rarely needs changes.

If you really don't like it, then I'll go for option 3. It's a kind of way to tell to the developer "You don't want the easiness provided by Docker ? Fine, install stuffs yourself.". This repo is already opinionated: by using Shell scripts you discourage the use of Windows. So I wouldn't be afraid to discourage the avoidance of Docker as well.

A last side-remark about this: if we lead the developer to have its JDK only in Docker instead of locally, he would need to tell his IDE to find its JDK in Docker, which is not straightforward (I haven't done it yet).

File: scaffolds/java/gradle.sh [hunk 10]
```diff
@@ -4,4 +4,4 @@ docker run --rm \
   -u $(id -u):$(id -g) \
   -v $(pwd):/home/gradle/project:rw \
   -w /home/gradle/project \
-  gradle:4.10-jdk10-slim gradle $@
+  gradle:7.4.2-jdk11-alpine gradle $@
```
[the remark this answers is not in the document]
Comment (ThomasCollignon) [reply scaffolds/java/gradle.sh:7]: That would depend on the choice we make for the previous comment: `gradle wrapper` fails if there is a file named `gradle` in the same folder as `gradle.build`. No issue if we use `gradle.sh`.

File: scaffolds/java/scripts/build (hunk 8 above)
Comment (ThomasCollignon) [reply scaffolds/java/scripts/build:5]: You're implying that if a developer wants to use Docker, then he has no `gradlew` nor Gradle installed, and I think it's wrong.
Developers may firstly use `gradlew`, then for any reason, may want to give a try with Docker, then the `build|test` scripts will still use `gradlew`. And I don't want to tell the developer that using Docker and having `gradlew` are mutually exclusive.

Comment (jbruggem) [reply scaffolds/java/scripts/build:5]: > You're implying that if a developer wants to use Docker, then he has no gradlew nor Gradle installed, and I think it's wrong.

I'm not implying anything :-). Just to clarify: the reason I suggest this script is that it would be a great way to make the TCR scripts work automatically with whatever the user has available to run the tests.

In any case, it seems highly probable to me, that someone that has installed gradle or a gradle wrapper did it with the intent to use it. The docker wrapper, in my view, is a fallback for those who do not wish to install the development environment for a given language. 

> And I don't want to tell the developer that using Docker and having gradlew are mutually exclusive.

I don't believe that's what we're saying here, and I don't believe this script makes them mutually exclusive: you're always free to run `./gradlew` or a `docker run ...` yourself. The script is not here to cover all usages (that would be unrealistic), but to cover the most common ones.

The objective of these scaffolds is to get started quickly, without any fuss. The point is to go from zero to coding very quickly, whether a dev env is present on your machine or not. Everything that is handled automatically is a plus. Every choice that must be made is friction we want to avoid.

If a dev has already taken the time to install a `./gradlew`, it's a good indicator that this is how they would like to work for this repo. If it's not any more, they can delete the wrapper or call the docker image directly.

File: scaffolds/java/gradle/wrapper/gradle-wrapper.properties (hunk 9 above)
[the remark this answers is not in the document]
Comment (jbruggem) [reply scaffolds/java/gradle/wrapper/gradle-wrapper.properties:1]: Thanks for this well-structured and insightful answer :). 

> Which means that it requires to have a Gradle version already installed, 

Indeed. This is consistent with the fact that you need an `npm` for Javascript, `cargo` for Rust and  `mix` for Elixir -- so it seems quite normal to me.

> and if it doesn't match the syntax in build.gradle, then he's screwed.

Though this *can* happen, it's quite improbable here with such a simple `build.gradle`. If it happens though, you can simply briefly rename `build.gradle` and do `gradle wrapper --gradle-version X.X`. Probably worth a line in the README though, it explain that there's an "escape hatch".

> If you really don't like it, then I'll go for option 3. It's a kind of way to tell to the developer "You don't want the easiness provided by Docker ? Fine, install stuffs yourself.".

It seems to me indeed inescapable that somebody not using docker will need to install some stuff themselves (for example: the JVM).  For me, committing the wrapper version in the `build.gradle` is more in line how dependency management is always done: don't provide the binary, but commit the version needed so that it can be downloaded.

> This repo is already opinionated: by using Shell scripts you discourage the use of Windows. 

Genuine question: in what way does providing scripts for Linux discourage the use of Windows ? 

To be clear, a PR that adds support for Windows is more than welcome. We started from our needs, not from a will to exclude anybody. I don't personally use Windows, so I'm not sure what would be needed to support Windows. If we can support both, that would be great of course !

> A last side-remark about this: if we lead the developer to have its JDK only in Docker instead of locally, he would need to tell his IDE to find its JDK in Docker, which is not straightforward (I haven't done it yet).

We included a docker wrapper for users without a proper Java IDE and that kind of stuff. We assumed that users with an IDE such as Eclipse/Intellij usually already have all they need to use gradle directly. 

We were more aiming at users that have a lightweight code editor and don't usually code in Java.

[the remark this answers is not in the document]
Comment (ThomasCollignon) [reply scaffolds/java/gradle/wrapper/gradle-wrapper.properties:1]: > Though this can happen, it's quite improbable here with such a simple build.gradle.

It currently happens to all developers using Gradle 7+ and the master of this scaffold because of `testCompile`. But ok, I agree that we should now be safe for a couple of years.

I'll go for option 3 then.

> Genuine question: in what way does providing scripts for Linux discourage the use of Windows ?

AFAIK Windows doesn't natively support the execution of Shell scripts, a tool like Cygwin or Git Bash is required.

File: scaffolds/java/gradle.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (ThomasCollignon) [reply scaffolds/java/gradle.sh:7]: ok

File: scaffolds/java/scripts/build (hunk 8 above)
Comment (ThomasCollignon) [reply scaffolds/java/scripts/build:5]: > Every choice that must be made is friction we want to avoid.

Ok I get your point. You have a practical point of view, while I have a more pedagogic one.

At the end, this repo should follow a single vision. I'll add you script.

**Review by jbruggem — APPROVED** [review]

**Comment by jbruggem** [comment]
@ThomasCollignon many thanks for your contribution *and* your patience :).
